File: algorithms/T/Tuple with Same Product/Tuple with Same Product.java
class Solution {
    public int tupleSameProduct(int[] nums) {
        
        if(nums.length < 4){
            return 0;
        }
        
        int res = 0;
        
        HashMap<Integer, Integer> map = new HashMap<>();
        
        for(int i = 0; i < nums.length - 1; i++){
            
            for(int j = i + 1; j < nums.length; j++){
                
                int val = nums[i] * nums[j];
                map.put(val, map.getOrDefault(val, 0) + 1);
            }
        }
        
        for(int key : map.keySet()){
            
            int val = map.get(key);
            
            if(val > 1){
                res += val * (val - 1) * 4;    // (val * (val - 1) / 2) * 8
            }
        }
        
        return res;
    }
}

File: scripts/algorithms/T/Tuple with Same Product/Tuple with Same Product.cpp
class Solution {
public:
    int tupleSameProduct(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int,int> map;
        int res = 0;
        for(int i = 0 ; i < n ; i++){
            for(int j = i+1 ; j < n ; j++){
                int prod = nums[i] * nums[j];
                map[prod]++;//store product of each possible pair
            }
        }
        for(pair<int,int> m:map){
            int n=m.second;
            res += (n*(n-1))/2; //no. of tuple
        }
        return res*8; //Every tuple has 8 permutations
    }
};
